[CheckpointWrapper] Warn on reentrant use

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102890
* #102692
* #102680
* #102672
* #102658
* #102657
* #102639
* #102637
* #102508

We'd like to encourage users to try non-reentrant as much as possible,
and identify any gaps this way.

Differential Revision: [D46397786](https://our.internmc.facebook.com/intern/diff/D46397786/)

rohan-varma committed:
[CheckpointWrapper] Warn on reentrant use

We'd like to encourage users to try non-reentrant as much as possible,
and identify any gaps this way.

Differential Revision: [D46397786](https://our.internmc.facebook.com/intern/diff/D46397786/)

[ghstack-poisoned]

awgu reviewed (approved):
I think this is duplicated with https://github.com/pytorch/pytorch/pull/102850?

Also, reminder to address https://github.com/pytorch/pytorch/pull/102850#discussion_r1214657220 before landing

rohan-varma committed:
Update on "[CheckpointWrapper] Warn on reentrant use"


We'd like to encourage users to try non-reentrant as much as possible,
and identify any gaps this way.

Differential Revision: [D46397786](https://our.internmc.facebook.com/intern/diff/D46397786/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[CheckpointWrapper] Warn on reentrant use"


We'd like to encourage users to try non-reentrant as much as possible,
and identify any gaps this way.

Differential Revision: [D46397786](https://our.internmc.facebook.com/intern/diff/D46397786/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
